src/storage.zig

```diff
@@ -69,77 +71,178 @@ pub const Storage = struct {
     };
 
     pub const NextTick = struct {
-        next: ?*NextTick = null,
+        next: ?*NextTick,
         callback: fn (next_tick: *NextTick) void,
     };
 
+    const ThreadPool = struct {
+        mutex: std.Thread.Mutex = .{},
+        cond: std.Thread.Condition = .{},
+        queue: FIFO(NextTick) = .{ .name = "thread-pool" },
+        notified: bool = false,
+        idle: u16 = 0,
+
+        var instance = ThreadPool{};
+        var init_once = std.once(ThreadPool.init);
+        var init_result: std.Thread.SpawnError!void = undefined;
+
+        fn init() void {
+            init_result = spawn();
+        }
+
+        fn spawn() std.Thread.SpawnError!void {
+            var threads = @maximum(1, std.Thread.getCpuCount() catch 1);
+            while (threads > 0) : (threads -= 1) {
+                const thread = try std.Thread.spawn(.{}, poll, .{&instance});
+                thread.detach();
+            }
+        }
+
+        fn submit(pool: *ThreadPool, next_tick: *NextTick) void {
+            pool.mutex.lock();
+            pool.queue.push(next_tick);
+            pool.unlock_after_queue_update();
+        }
+
+        fn poll(pool: *ThreadPool) void {
+            pool.mutex.lock();
+            while (true) {
+                while (pool.queue.pop()) |next_tick| {
+                    pool.unlock_after_queue_update();
+                    next_tick.callback(next_tick);
+                    pool.mutex.lock();
+                }
+
+                pool.idle += 1;
+                pool.cond.wait(&pool.mutex);
+                pool.idle -= 1;
+                pool.notified = false;
+            }
+        }
+
+        fn unlock_after_queue_update(pool: *ThreadPool) void {
+            if (!pool.queue.empty() and !pool.notified and pool.idle > 0) {
+                pool.notified = true;
+                pool.cond.signal();
+            }
+            pool.mutex.unlock();
+        }
+    };
+
     io: *IO,
     fd: os.fd_t,
+    main_thread_id: std.Thread.Id,
```

Comment:
Can you use this for `assert(storage.is_background_thread())` in callbacks that are run in the threadpool (eg `cpu_sort_callback`), to call out that they need to be read carefully.

Reply:
This would need to be exposed/work on `testing/storage.zig` too (as Tree is used on both Storage types). But test storage doesn't use threads.

Reply:
Good point. We could have testing/storage store a boolean flag that's true when it's executing a background callback?

Just thinking ahead to when I inevitably mess up thread safety :)